### PR TITLE
Add support for streaming rendered components using `renderToPipeableStream`

### PR DESCRIPTION
**Pull request by AbanoubGhadban**
## Summary
This PR introduces the new features introduced in React 18 which is supporting [Async Components](https://github.com/shakacode/react_on_rails_pro/pull/407/files#diff-10afc60eb22f2d926f16455d04c3cb13038e9e6fd8492e966d01dd8b7fa0c731R9) and streaming components. It's done by replacing [renderToString](https://github.com/shakacode/react_on_rails/pull/1633/files#diff-cbb0f0518343f42b78a0ec373eddd1187ffb9275651ddb9758a1160c52176eb6R74) function with [renderToPipeableStream](https://github.com/shakacode/react_on_rails/pull/1633/files#diff-cbb0f0518343f42b78a0ec373eddd1187ffb9275651ddb9758a1160c52176eb6R204).
https://www.loom.com/share/db55036514b746a7b283e6f296134168?sid=2ce8205d-6956-4e9a-97bc-ee55004f3216

## The Dummy App
To open the newly introduced page in the dummy app, you need to clone both **React on Rails** and **React on Rails Pro** on your local machine and put them in the same directory
- <parent_directory>
  - react_on_rails
  - react_on_rails_pro

execute the following commands
```
cd <parent_directory>
cd react_on_rails
bundle && yarn
cd ../react_on_rails_pro
bundle && yarn
cd spec/dummy
bundle && yarn

bin/dev
```

Then, open the newly introduced `stream_async_components` in the dummy app.

## More resources
https://react.dev/reference/react-dom/server/renderToPipeableStream
https://github.com/reactwg/react-18/discussions/37

## Sequence Diagrams

```mermaid
sequenceDiagram
    participant Browser
    participant RailsServer
    participant ReactOnRails (Ruby)
    participant ReactOnRailsPro (JS)
    participant ReactDOMServer
    Browser->>RailsServer: Request page with React component
    RailsServer->>ReactOnRails (Ruby): Call `stream_react_component`
    ReactOnRails (Ruby)->>ReactOnRailsPro (JS): Trigger `streamServerRenderedReactComponent`
    ReactOnRails (JS)->>ReactDOMServer: Call `renderToPipeableStream`
    ReactDOMServer-->>ReactOnRailsPro (JS): PipeableStream
    ReactOnRails (JS)-->>ReactOnRails (Ruby): Stream content
    ReactOnRails (Ruby)-->>RailsServer: Streamed React component content
    RailsServer-->>Browser: Serve content
    Note right of Browser: Displays page with streamed React component
```

## Summary by CodeRabbit

- **New Features**
	- Introduced streaming functionality for React components, enabling server-side rendering with enhanced performance and better error management.
	- Added methods for asynchronous streaming of React components and improved output formatting in the React on Rails helper module.
	- Implemented new functions to manage server-side rendering of React components to a stream.
	- Enhanced Webpack configuration to support Node.js globals and added new aliases for better compatibility.
	- Introduced a new method for accessing streaming options in the rendering process.
	- Added a comprehensive guide on implementing streaming server rendering using React 18's APIs.

- **Bug Fixes**
	- Improved error handling during rendering for increased reliability.
	- Clarified method name for better understanding in the React on Rails module.

- **Updates**
	- Upgraded React and React-DOM versions to 18.2.0 for improved support and additional features.
	- Updated test setup to ensure compatibility with jsdom environment, including a new polyfill for compatibility issues.
	- Improved version control cleanliness by ignoring IDE-related files and directories.
	- Updated documentation structure and links to reflect recent changes and enhancements.

- - -
This change is [Reviewable](https://reviewable.io/reviews/shakacode/react_on_rails/1633)

### CODE REVIEW

File: lib/react_on_rails/helper.rb [hunk 1]
```diff
@@ -91,6 +91,36 @@ def react_component(component_name, options = {})
...
+      options = options.merge(stream?: true)
+      result = internal_react_component(component_name, options)
+      build_react_component_result_for_server_streamed_content(
+        rendered_html_stream: result[:result],
+        component_specification_tag: result[:tag],
+        render_options: result[:render_options]
+      )
+    end
+
+    def stream_react_component_async(component_name, options = {})
```
Comment (AbanoubGhadban) [inline lib/react_on_rails/helper.rb]: To see how this function is used look [here](https://github.com/shakacode/react_on_rails_pro/pull/407/files#diff-7bbea53e72fefe60558edaf0ec093a01425ed7403fc60901320a7c3d9e154f8dR1) and [here](https://github.com/shakacode/react_on_rails_pro/pull/407/files#diff-4255e4faea72596818e5393cfddc705c9f7a72b4b1a77fef16b8835cfa387216R31)

File: lib/react_on_rails/helper.rb [hunk 2]
```diff
@@ -91,6 +91,36 @@ def react_component(component_name, options = {})
       end
     end
 
+    def stream_react_component(component_name, options = {})
+      options = options.merge(stream?: true)
+      result = internal_react_component(component_name, options)
```
Comment (AbanoubGhadban) [inline lib/react_on_rails/helper.rb]: The result is of type [StreamRequest](https://github.com/shakacode/react_on_rails_pro/pull/407/files#diff-c1d07594e0987e090d78fe43f67ee57bd751c72c4e5cec3df2ff04a8d2606c14R58)

File: lib/react_on_rails/helper.rb [hunk 3]
```diff
@@ -361,6 +391,32 @@ def build_react_component_result_for_server_rendered_string(
...
+      rendered_html_stream: required("rendered_html_stream"),
+      component_specification_tag: required("component_specification_tag"),
+      render_options: required("render_options")
+    )
+      content_tag_options_html_tag = render_options.html_options[:tag] || "div"
+      # The component_specification_tag is appended to the first chunk
+      # We need to pass it early with the first chunk because it's needed in hydration
+      # We need to make sure that client can hydrate the app early even before all components are streamed
+      is_first_chunk = true
+      rendered_html_stream = rendered_html_stream.transform do |chunk|
```
Comment (AbanoubGhadban) [inline lib/react_on_rails/helper.rb:442]: The [StreamRequest](https://github.com/shakacode/react_on_rails_pro/pull/407/files#diff-c1d07594e0987e090d78fe43f67ee57bd751c72c4e5cec3df2ff04a8d2606c14R58) and [StreamDecorator](https://github.com/shakacode/react_on_rails_pro/pull/407/files#diff-c1d07594e0987e090d78fe43f67ee57bd751c72c4e5cec3df2ff04a8d2606c14R2) classes allow us to transform the stream of http body chunks

File: lib/react_on_rails/helper.rb [hunk 4]
```diff
@@ -512,6 +570,9 @@ def server_rendered_react_component(render_options)
                                                js_code: js_code)
       end
 
+      # TODO: handle errors for streams
+      return result if render_options.stream?
```
Comment (AbanoubGhadban) [inline lib/react_on_rails/helper.rb:621]: The ability to show the error details happen while rendering a streamable component will be handled in a follow up PR

File: package.json [hunk 5]
```diff
@@ -39,8 +39,8 @@
     "prettier": "^2.8.8",
     "prettier-eslint-cli": "^5.0.0",
     "prop-types": "^15.8.1",
-    "react": "^17.0.0",
-    "react-dom": "^17.0.0",
+    "react": "^18.2.0",
```
Comment (AbanoubGhadban) [inline package.json:42]: `renderToReadableStream` function is added in React 18

**Review by AbanoubGhadban — COMMENTED** [review]
Self-reviewed

File: lib/react_on_rails/helper.rb [hunk 6]
```diff
@@ -404,20 +460,22 @@ def compose_react_component_html_with_spec_and_console(component_specification_t
...
+      content_tag(:script,
+                  json_safe_and_pretty(data).html_safe,
+                  type: "application/json",
+                  id: "js-react-on-rails-context")
+    end
 
-      "#{rails_context_content}\n#{render_value}".html_safe
+    # prepend the rails_context if not yet applied
+    def prepend_render_rails_context(render_value)
+      "#{rails_context_if_not_already_rendered}\n#{render_value}".html_safe
```
[the remark this answers is not in the document]
Comment (AbanoubGhadban) [reply lib/react_on_rails/helper.rb]: Used String#strip function to remove white spaces from the beginning of the content

File: node_package/tests/ReactOnRails.test.js [hunk 7]
```diff
@@ -19,10 +19,15 @@ describe('ReactOnRails', () => {
     });
     ReactOnRails.register({ R1 });
 
-    document.body.innerHTML = '<div id="root"></div>';
-    // eslint-disable-next-line no-underscore-dangle
-    const actual = ReactOnRails.render('R1', {}, 'root')._reactInternals.type;
```
Comment (AbanoubGhadban) [inline node_package/tests/ReactOnRails.test.js:24]: When we use a newer react version that supports the `Root API`, it returns the `ReactDomRoot` object, which doesn't have the `_reactInternals` property.

File: node_package/src/serverRenderReactComponent.ts [hunk 8]
```diff
@@ -165,4 +167,57 @@ const serverRenderReactComponent: typeof serverRenderReactComponentInternal = (o
...
+      props,
+      railsContext,
+    });
+
+    if (isServerRenderHash(reactRenderingResult) || isPromise(reactRenderingResult)) {
+      throw new Error('Server rendering of streams is not supported for server render hashes or promises.');
+    }
+
+    renderResult = new PassThrough();
+    ReactDOMServer.renderToPipeableStream(reactRenderingResult as ReactElement).pipe(renderResult);
```
[the remark this answers is not in the document]
Comment (AbanoubGhadban) [reply node_package/src/serverRenderReactComponent.ts]: You are right. The cast is not necessary.

**Comment by AbanoubGhadban** [comment]
```
Would it make sense to add async: false parameter to def stream_react_component instead of a separate function (probably not, but asking just in case)?
```

The `stream_react_component_async` is renamed now to `stream_react_component`. It's  async by default. The other function is removed

File: lib/react_on_rails/helper.rb [hunk 9]
```diff
@@ -361,6 +391,32 @@ def build_react_component_result_for_server_rendered_string(
       prepend_render_rails_context(result)
     end
 
+    def build_react_component_result_for_server_streamed_content(
+      rendered_html_stream: required("rendered_html_stream"),
+      component_specification_tag: required("component_specification_tag"),
+      render_options: required("render_options")
```
[the remark this answers is not in the document]
Comment (AbanoubGhadban) [reply lib/react_on_rails/helper.rb:435]: This way of throwing errors while not providing a required parameter is used in many places. I can't change it in this PR, we can change it in another PR.

File: .gitignore [hunk 10]
```diff
@@ -32,3 +32,6 @@ npm-debug.*
 yalc.lock
 
 .byebug_history
+
+# IDE
+.idea/
```
Comment (justin808) [inline .gitignore]: this is OK, but this should go in your gitignore_global for all projects

File: lib/react_on_rails/helper.rb [hunk 11]
```diff
@@ -330,6 +352,16 @@ def load_pack_for_generated_component(react_component_name, render_options)
 
     private
 
+    def stream_react_component_internal(component_name, options = {})
```
[the remark this answers is not in the document]
Comment (justin808) [reply lib/react_on_rails/helper.rb]: @AbanoubGhadban I agree.

File: lib/react_on_rails/helper.rb [hunk 12]
```diff
@@ -361,6 +393,33 @@ def build_react_component_result_for_server_rendered_string(
...
+            <#{content_tag_options_html_tag} id="#{render_options.dom_id}">#{chunk}</#{content_tag_options_html_tag}>
+          HTML
+          html_content.strip
+        end
+        chunk
+      end
+
+      rendered_html_stream.transform(&:html_safe)
+      # TODO: handle console logs
+    end
```
[the remark this answers is not in the document]
Comment (justin808) [reply lib/react_on_rails/helper.rb:457]: help

File: node_package/src/serverRenderReactComponent.ts [hunk 13]
```diff
@@ -165,4 +167,57 @@ const serverRenderReactComponent: typeof serverRenderReactComponentInternal = (o
...
+
+    // TODO: Add console replay script to the stream
+    // Ensure to avoid console messages leaking between different components rendering
```
[the remark this answers is not in the document]
Comment (justin808) [reply node_package/src/serverRenderReactComponent.ts]: implement

File: node_package/src/serverRenderReactComponent.ts [hunk 14]
```diff
@@ -165,4 +166,57 @@ const serverRenderReactComponent: typeof serverRenderReactComponentInternal = (o
...
+    if (isServerRenderHash(reactRenderingResult) || isPromise(reactRenderingResult)) {
+      throw new Error('Server rendering of streams is not supported for server render hashes or promises.');
+    }
+
+    renderResult = new PassThrough();
+    ReactDOMServer.renderToPipeableStream(reactRenderingResult).pipe(renderResult);
+
+    // TODO: Add console replay script to the stream
+    // Ensure to avoid console messages leaking between different components rendering
+  } catch (e: any) {
```
[the remark this answers is not in the document]
Comment (justin808) [reply node_package/src/serverRenderReactComponent.ts]: implement

**Comment by justin808** [comment]
@AbanoubGhadban let's see if this is ready for merge and pre-release.

File: node_package/src/serverRenderReactComponent.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (AbanoubGhadban) [reply node_package/src/serverRenderReactComponent.ts]: Implemented in #1647

**Pull request merged**